modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.ingest.ConfigurationUtils.newConfigurationException;
+
+public final class DataStreamRouterProcessor extends AbstractProcessor {
+    public static final String TYPE = "data_stream_router";
+
+    private static final String DATA_STREAM_PREFIX = "data_stream.";
+    private static final String DATA_STREAM_TYPE = DATA_STREAM_PREFIX + "type";
+    private static final String DATA_STREAM_DATASET = DATA_STREAM_PREFIX + "dataset";
+    private static final String DATA_STREAM_NAMESPACE = DATA_STREAM_PREFIX + "namespace";
+    private static final String DATA_STREAM_DEFAULT_DATASET = "generic";
+    private static final String DATA_STREAM_DEFAULT_NAMESPACE = "default";
+    private static final String EVENT_DATASET = "event.dataset";
+
+    private static final char[] DISALLOWED_IN_DATASET = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':', '-' };
+    private static final char[] DISALLOWED_IN_NAMESPACE = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':' };
+    private static final int MAX_LENGTH = 100;
+    private static final char REPLACEMENT_CHAR = '_';
+    private final String type;
+    private final String dataset;
+    private final String namespace;
+
+    private DataStreamRouterProcessor(String tag, String description, String type, String dataset, String namespace) {
+        super(tag, description);
+        this.type = type;
+        this.dataset = dataset;
+        this.namespace = namespace;
+    }
+
+    private static String sanitizeDataStreamField(String s, char[] disallowedInDataset) {
+        if (s == null) {
+            return null;
+        }
+        s = s.toLowerCase(Locale.ROOT);
+        s = s.substring(0, Math.min(s.length(), MAX_LENGTH));
+        for (char c : disallowedInDataset) {
+            s = s.replace(c, REPLACEMENT_CHAR);
```

Comment:
I don't like the magical replacement. Ca we just make it invalid and return an error?

Reply:
When you configure the processor with an invalid value, it will return an error.

However, when taking values from the document, it will be more lenient. One use case for the leniency is the fallback to `event.dataset`. ECS logger have first-class support to setting the value as it's an important field for the logs UI which shows log rates per `event.dataset`. The `event.dataset` field did not have restrictions on the value previously. Introducing them retrospectively would be a breaking change. Therefore, I opted to sanitize rather than reject invalid values. See also the [ECS RFC about data streams](https://github.com/elastic/ecs/blob/main/rfcs/text/0009-data_stream-fields.md#scope-of-impact):

> Since data_stream.dataset and event.dataset should contain the same value, the restrictions imposed on data_stream.dataset might affect the event.dataset value. This means users may need to translate their custom dataset values (e.g. event.dataset: firewall/config) to an equivalent legal dataset, according to the character restrictions imposed by the use of the value in data_stream.dataset, for example data_stream.dataset: firewall.config.

We could change it so that only `event.dataset` values will be sanitized. But completely dropping an event if the target contains an invalid character seems somewhat harsh, especially since we don't have a dead letter queue.
Do we really prefer to completely drop a log event if it contains `"data_stream.dataset": "my-service"`? To me, sanitizing the value from `my-service` to `my_service` seems more user-friendly.

But I'm open to changing this if you still feel strongly about it.

Reply:
> So building event.dataset into more places will make it even harder to get rid off. Could we skip event.dataset as a fallback in a first iteration?

As it is already built in into ECS loggers and the logs UI, not having support for it would cause challenges. But we could add a flag to the processor (such as `event_dataset_fallback`) that's off by default.

Reply:
++

Reply:
Removed any usage of `event.dataset`, as discussed. If fallbacks are desired, a `set` processor can be used.